File: codes/31. 下一个排列.go
func nextPermutation(nums []int) {
	n := len(nums)
	pivot := n - 1
	for pivot > 0 && nums[pivot] <= nums[pivot-1] {
		pivot--
	}
	if pivot == 0 {
		// the last permutation
		for i := 0; i < n/2; i++ {
			nums[i], nums[n-1-i] = nums[n-1-i], nums[i]
		}
		return
	}
	pivot--
	nextPivotNum := math.MaxInt32
	nextPivot := -1
	for i := pivot + 1; i < n; i++ {
		if nums[i] > nums[pivot] && nums[i] <= nextPivotNum {
			nextPivotNum = nums[i]
			nextPivot = i
		}
	}
	nums[pivot], nums[nextPivot] = nums[nextPivot], nums[pivot]
	remainLen := n - pivot - 1
	remainNums := nums[pivot+1:]
	for i := 0; i < remainLen/2; i++ {
		remainNums[i], remainNums[remainLen-1-i] = remainNums[remainLen-1-i], remainNums[i]
	}
}

File: codes/31. 下一个排列.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        auto n = nums.size();
        if(n<=1) {
            return;
        }
        int begin = n-2;
        for(; begin>=0; begin--) {
            if(nums[begin] >= nums[begin+1]) {
                continue;
            }
            break;
        }
        sort(nums.begin()+begin+1, nums.end());
        if(begin==-1) {
            return;
        }
        auto pivot = nums[begin];
        int idx = upper_bound(nums.begin()+begin+1, nums.end(), pivot)-nums.begin();
        nums[begin] = nums[idx];
        nums[idx] = pivot;
    }
};